Resolves #17 - Run history / seed manager

MattEqualsCoder opened the pull request.

Vivelin commented on src/Randomizer.Shared/Models/Seed.cs:
```diff
@@ -1,19 +1,21 @@
-    public class Seed {
+namespace Randomizer.Shared.Models
+{
+    public class Seed
```
I know you just repurposed this file but I think we can come up with a better name than "seed" because "seed" to me is the number used for the RNG, not the entire generated game. Maybe not an entirely different name, but at least enough to clearly differentiate between the generated game "seed" and the RNG seed

MattEqualsCoder replied:
For our purposes, probably something like "GeneratedRom" or something like that would be more accurate?

Vivelin replied:
Yeah, I think that would work

Vivelin commented on src/Randomizer.Shared/Models/Seed.cs:
```diff
@@ -1,19 +1,21 @@
-        public int Players { get; set; }
-        public List<World> Worlds { get; set; }
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public DateTime Date { get; set; }
```
Should probably be DateTimeOffset, and I don't know if having the date as ID is a good idea. Maybe just stick to an int ID for this?

Vivelin commented on src/Randomizer.Shared/Models/TrackerDungeonState.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public int Id { get; set; }
+        public int TrackerStateId { get; set; }
```
What is this value?

MattEqualsCoder replied:
Just foreign key to TrackerState object.

Vivelin replied:
Oh. Please add a navigation property (`public TrackerState TrackerState {get; set;}`) to clear that up, then. Also makes it easier to use.

MattEqualsCoder commented on src/Randomizer.Shared/Models/TrackerDungeonState.cs:
```diff
@@ -0,0 +1,18 @@
+        public string Name { get; set; }
+        public bool Cleared { get; set; }
+        public int RemainingTreasure { get; set; }
+        public int Reward { get; set; }
```
I'll have to review the usage of the Shared project in the SMZ3 project. I initially tried to do that, but it didn't let me due to the SMZ3 project not being a dependency, and I can't introduce it as a dependency as it would create a cyclic dependency. But I think with all of this being managed by the App I can probably pull out all references to the Shared project from the SMZ3 one now.

Vivelin replied:
Yeah, some of the basic enums that are used everywhere could probably be moved to there

Vivelin commented on src/Randomizer.SMZ3/.cs:
```diff
@@ -0,0 +1,30 @@
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Field,
+        Inherited = false, AllowMultiple = false)]
+    public sealed class ItemCategoryAttribute : Attribute
```
This file doesn't have a file name?

MattEqualsCoder replied:
Ah, think I accidentally deleted the filename rather than the file. lol

Vivelin commented:
When you start the randomizer for the first time after the update, the screen is empty. In this case, I feel like it might be nicer to at the very least show a message there in the center. Maybe something along the lines of "You haven't started any games yet. Click here to create a new game", and then add a copy of the "Custom generate" button?

To test this, I imagine we need a way to delete generated roms from this list, or clearing it entirely.

I feel like "Quick generate" should be changed back into "Play" and also launch the generated game, considering there is not much else you can do besides start it at this point.

"Custom generate" should be renamed to "New..." or "Custom..." or maybe even "Generate custom game...". The `...` is used within Windows to indicate a dialog with more options opens up before the associated action is really executed.

In the window that opens, I would suggest swapping the "Generate" and "Cancel" buttons to be more consistent with Windows UI.

Finally, when I hit "Launch", I get a message saying that no speech recognition capabilities were detected, which is obviously wrong. It looks like `CanStartTracker` is never set to `true` anywhere?

Otherwise, great work so far! This really does add a lot of value, I think.

MattEqualsCoder commented:
Implemented all of these changes. I had meant to actually add a little intro thing when there were no roms, but forgot, so thanks for the reminder on that one!

Sorry for the bug. That's what I get for making last minute updates!

I have added comments and did some code clean up. So unless you have any other concerns or changes, I think this should be fully ready to go. I did some testing with adding, deleting, and renaming seeds and played a bit to make sure everything still ran and tracked things as expected.

Vivelin commented on src/Randomizer.App/RomGenerator.cs:
```diff
@@ -78,12 +89,24 @@ protected byte[] GenerateRomBytes(RandomizerOptions options, out SeedData seed)
             return _randomizer.GenerateSeed(config, seed ?? options.SeedOptions.Seed, CancellationToken.None);
         }
 
+        /// <summary>
```
Just as an FYI, you don't really need to add documentation to private or internal stuff, unless you think it's necessary to understand it properly.

Vivelin commented on src/Randomizer.SMZ3.Tracking/Tracker.cs:
```diff
@@ -1225,6 +1258,7 @@ public void MarkDungeonAsCleared(DungeonInfo dungeon, float? confidence = null)
         /// <param name="confidence">The speech recognition confidence.</param>
         public void MarkDungeonAsIncomplete(DungeonInfo dungeon, float? confidence = null)
         {
+            dungeon = Dungeons.First(x => x.Name == dungeon.Name);
```
What's the point of this?

Vivelin reviewed (changes requested):
I don't know if this was something you changed on purpose or something that got lost when merging main back into this branch, but why did you move the Dungeons back into Tracker and introduce Clone()? 

I'm not sure why Clone would be necessary, and I purposefully moved the Dungeons out of tracker into the WorldInfo property with all the other configurable location stuff. I'm not sure about some of the lines you added.

MattEqualsCoder commented:
Perhaps separate classes should be created for tracker tracking the state of things, then? Or maybe it should use the DB classes?

The problem comes from the fact that each tracker pulls their initial information from config factories, meaning that there is only a single instance of these starting config classes that ends up getting shared amongst multiple instances of the tracker.

Basically, if you were to launch a new window of tracker for a brand new rom after closing out a tracker window that had some data tracked, that data would transfer over. The cloning was to create new instances of those objects to avoid that.

The core issue is that these config files contain some of the properties used for tracking the state of the tracker. So yeah, I can probably switch over to using those db shared classes for that stuff and we only use the json classes for things like names?

Vivelin commented:
Wouldn't changing the DI scope to Scoped also fix the issue, then? Or simply removing the pre-loaded instances and only injecting the config provider? That way, you have new, clean instances every time you start Tracker again.

MattEqualsCoder commented:
I suppose part of me was wanting to avoid reloading the json files each time, but admittedly I'm not quite sure why. (I also wasn't fully aware of AddScoped, and that indeed fixes it quite easily.) Changes have been made.

Vivelin commented on src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs:
```diff
@@ -33,18 +33,18 @@ public static IServiceCollection AddTracker<TWorldAccessor>(this IServiceCollect
             services.AddSingleton<IWorldAccessor>(x => x.GetRequiredService<TWorldAccessor>());
-            services.AddSingleton<TrackerConfigProvider>();
-            services.AddSingleton(serviceProvider =>
+            services.AddScoped<TrackerConfigProvider>();
```
The provider itself can stay singleton, as it doesn't cache the instances it creates. Not too important unless we need it somewhere that doesn't use a service scope.

MattEqualsCoder replied:
Updated
